proposals/route-blinding.md

```diff
@@ -0,0 +1,607 @@
+```
+
+Alice creates a blinded route `Carol -> Bob -> Alice`.
+Alice has chosen what fee settings will be used inside the blinded route.
+Let's assume she has chosen `fee_base_msat = 10` and `fee_proportional_millionths = 100`.
+
+The attacker knows that the recipient is at most two hops away from Carol. Instead of making the
+payment, the attacker watches for new `channel_update`s for every channel in a two-hops radius
+around Carol. At some point, the attacker sees a `channel_update` for the channel `Bob -> Alice`
+that sets `fee_proportional_millionths = 150`, which exceeds what Alice has chosen for the blinded
+route. The attacker then tries to make the payment.
+
+When Bob receives the payment, the fees are below its current settings, so it should reject it.
+The attacker would then receive a failure, and be able to infer that it's very likely that Alice
+is the final recipient.
+
+If the attackers are able to frequently request invoices from the recipient (e.g. from a Bolt 12
+offer), they don't even have to attempt the payment to detect this. They can simply periodically
+request invoices from the recipient and detect when the recipient raises the fees or cltv of the
+blinded route, and match that with recent `channel_update`s that they received.
+
+Similarly, feature bits that apply to payment relaying behavior can be used to fingerprint nodes
+inside the blinded route: this is why `allowed_features` are committed inside the `encrypted_data`.
+
+If nodes across the network use different values for `htlc_minimum_msat`, it can also be used to
+fingerprint nodes: that's why it is also committed inside the `encrypted_data`.
+
+This type of attack is the reason why all parameters that affect payment relaying behavior (fees,
+cltv, features, etc) are chosen by the recipient. The recipient should add a large enough margin
+to the current values actually used by nodes inside the route to protect against future raises.
+This is also why blinded routes used for payments have a `max_cltv_expiry` set by the recipient,
+even though that doesn't fully address the issue if the attackers are able to frequently request
+new blinded routes.
+
+Altruistic relaying nodes inside a blinded route could choose to relay payments with fees below
+their current settings, which would break this heuristic: however their economic incentive is to
+reject them, so we cannot rely on them to protect recipient privacy.
+
+Similarly, we mandate relaying nodes to only accept payments using exactly the fees provided in
+the `encrypted_data` payload. Otherwise, when observing a `channel_update` that raises a specific
+channel's fees, the attackers could try to use these new fees in a payment attempt: if the payment
+goes through, they would have even more confidence about the channel used in the blinded route.
+The incentives for relaying nodes aren't great, because we're asking them to reject payments that
+give them the right amount of fees to protect recipient privacy.
```

Comment:
optional suggestion: Could be worth opening an issue to add a "grace period" for nodes to allow old fees etc, like I think @rustyrussell mentioned CLN does. If so, that could be a future direction that could be worth noting here.

Reply:
Agreed, eclair also provides a grace period where it accepts old fees. I'll open a PR to add this recommendation to the spec, and bring it to the attention of other implementations.

Reply:
I opened a PR for that in #1001